### ISSUE TITLE
Network gives NaN after 30k trainings?

### ISSUE THREAD

**Issue by PetarIvancevic** [BrainJS/brain.js#227]
![Motivating the network](https://media.giphy.com/media/Bp1urA9f0XrW/giphy.gif)

## *What* is wrong?
I am working on a neural network, which implements reinforcement learning to play tetris. At one point during the training process with the learning rate set to 0.3 the network after 30k trainings gives the result 0 and then goes to NaN. Tried to figure out why Nan, but can't seem to.

## *Where* does it happen?
During the training process, when running a network with:
```
50 - input neurons
700 - first hidden
200 - second hidden
1 - output
```

## *How* do we replicate the issue?
Just run the network with the [700, 200] neurons and learning rate of 0.3.

## *How* important is this (1-5)?
2

## Expected behavior (i.e. solution)
It should give at least 0 if nothing else

## Other Comments

The entire code is here https://github.com/PetarIvancevic/node-neural-tetris . The code is run with:
```
node index.js -g NUM_GAMES -n NAME_OF_FOLDER
```

Ou yeah..the network LEARNS how to play...but it gives NaN xD....just weird. The blocks are fixated to always be OBlocks for testing and comparison.

**Comment by mubaidr** [comment]
I have tested the provide code (with brain.js 1.2.4) and this is the result of network training: 

for some games we have:
```bash
{ error: 0.0007142284249853261, iterations: 1 }
```

and most have this:
```bash
{ error: 2.099387610681942e-12, iterations: 1 }
```

with this input:
```bash
node index.js -g 100 -n NAME_OF_FOLDER
```

Am I missing something?

**Comment by PetarIvancevic** [comment]
You should "play" around 30k games, then it breaks. My latest guess is that it might be somehow node related, but I have no idea how :/

**Comment by mubaidr** [comment]
Sure, will do some testing.

**Comment by nandihno** [comment]
I have notice something odd,
I have been receiving NAN and then i ran the network with log:true
I saw that the training error went to NAN.
digging a bit further i found that where sigmoid is calculated
specifically  where the weights and inputs are calculated, there would be occasions where the inputs array is much smaller.
so looking at the code you have
` var sum = this.biases[layer][node];
   for (var k = 0; k < weights.length; k++) {
          sum += weights[k] * input[k];
    }
     //sigmoid
    this.outputs[layer][node] = 1 / (1 + Math.exp(-sum));
`
So where the input array is of smaller length then you will get NAN for the sum calculation and eventually for the sigmoid calculation.
Is this meant to be like this? i though each input will have its own weight?
Thanks

**Comment by robertleeplummerjr** [comment]
I see your project uses leaky-relu: https://github.com/PetarIvancevic/node-neural-tetris/blame/6ed746f0d2c376e25a1171ccb02cc7b41475d7c1/config.js#L10

Have you tried using any different activation with the same results?

**Comment by robertleeplummerjr** [comment]
Coming back to this: `leakyReluAlpha` may help with this, it was added around 1.5.0: https://github.com/BrainJS/brain.js#activation

**Comment by PetarIvancevic** [comment]
Ty all for trying to fix this. I'll tryout the leakyReluAlpha over the weekend to see how it will behave. 
@nandihno The inputs "should" all be the same length. If they are not the same length then that is a bug 😂

**Comment by robertleeplummerjr** [comment]
Ty @PetarIvancevic, will be super excited to hear how this comes along.  I'd love to implement reusable functionality into brain.js if you are interested in partnering up for this, if you see something worthwhile that is.

**Comment by nandihno** [comment]
Npm updated to 1.5.0? 
Also will be cool if there was a cdn 😊

**Comment by robertleeplummerjr** [comment]
Yup: https://www.npmjs.com/package/brain.js
For CDN all one needs is to find one for either npm or github.  We have the browser.js file already compiled for browser when needed as well.
